enumerable_methods.rb

```diff
@@ -0,0 +1,240 @@
+# frozen_string_literal: true
+
+module Enumerable
+  # my_each
+  def my_each
+    return to_enum unless block_given?
+
+    if self.class == Array
+      i = 0
+      while i < length
+        yield(self[i])
+        i += 1
+      end
+    elsif self.class == Hash
+      keys = self.keys
+      keys.length.times do |item|
+        key = keys[item]
+        value = self[key]
+        yield(key, value)
+      end
+    end
+    self
+  end
+
+  # my_each_with_index
+  def my_each_with_index
+    return to_enum unless block_given?
+
+    if self.class == Array
+      i = 0
+      while i < length
+        yield(self[i], i)
+        i += 1
+      end
+    elsif self.class == Hash
+      keys = self.keys
+      keys.length.times do |item|
+        key = keys[item]
+        value = self[key]
+        key_value = [key, value]
+        yield(key_value, item)
+      end
+    end
+    self
+  end
+
+  # my_select
+  def my_select
+    return to_enum unless block_given?
+
+    new_arr = []
+    i = 0
+    while i < size
+      new_arr << self[i] if yield(self[i])
+      i += 1
+    end
+    new_arr
+  end
+
+  # my_all
+  # rubocop:disable Metrics/PerceivedComplexity,Metrics/CyclomaticComplexity
+  def my_all?(pattern = nil)
+    arr = self
+    if !block_given? && pattern.nil?
+      arr.my_each do |i|
+        return false unless i
+      end
+    elsif self.class == Array
+      arr.my_each do |i|
+        return false unless yield(self[i])
+      end
+    elsif self.class == Hash
+      keys = self.keys
+      arr.my_each do |i|
+        return false unless yield(keys[i], self[keys[i]])
+      end
+    elsif pattern.is_a? Regexp
+      arr.my_each do |i|
+        return false unless i =~ pattern
+      end
+    elsif pattern.is_a? Class
+      arr.my_each do |i|
+        return false unless i.is_a? pattern
+      end
+    elsif pattern
+      arr.my_each do |i|
+        return false unless i == pattern
+      end
+    end
+    true
+  end
+  # rubocop:enable Metrics/PerceivedComplexity,Metrics/CyclomaticComplexity
+
+  # my_any
+  # rubocop:disable Metrics/PerceivedComplexity,Metrics/CyclomaticComplexity
+  def my_any?(pattern = nil)
+    arr = self
+    if !block_given? && pattern.nil?
+      arr.my_each do |i|
+        return true if i
+      end
+    elsif self.class == Array
+      arr.my_each do |i|
+        return true if yield(self[i])
+      end
+    elsif self.class == Hash
+      keys = self.keys
+      arr.my_each do |i|
+        return true if yield(keys[i], self[keys[i]])
+      end
+    elsif pattern.is_a? Regexp
+      arr.my_each do |i|
+        return true if i =~ pattern
+      end
+    elsif pattern.is_a? Class
+      arr.my_each do |i|
+        return true if i.is_a? pattern
+      end
+    elsif pattern
+      arr.my_each do |i|
+        return true if i == pattern
+      end
+    end
+    false
+  end
```

Comment:
- [ ] when a class is passed as an argument returns true if at least one of the collection is a member of such class
- [ ] when a Regex is passed as an argument returns false if none of the collection matches the Regex
- [ ] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern